src/core/algorithms/od/fastod/stripped_partition_cache.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <memory>
+
+#include "attribute_pair.h"
+#include "attribute_set.h"
+#include "cache_with_limit.h"
+#include "complex_stripped_partition.h"
+
+namespace algos::fastod {
+
+class StrippedPartitionCache {
+private:
+    CacheWithLimit<AttributeSet, ComplexStrippedPartition> cache_{static_cast<size_t>(1e8)};
+
+public:
+    void Clear() {
+        cache_.Clear();
+    }
+
+    ComplexStrippedPartition GetStrippedPartition(AttributeSet const& attribute_set,
+                                                  DataFrame const& data) {
+        if (cache_.Contains(attribute_set)) {
+            return cache_.Get(attribute_set);
+        }
+
+        std::optional<ComplexStrippedPartition> result;
```

Comment:
Personally I don't like use of optional here, because optional is indented to be used when the absence of a value is possible and expected. Here we use it to indicate that `result` was not initialized yet so that we can initialize it later, at the end `result` always has a value. Such a usage IMHO may confuse the reader.
The more readable way IMO is to use a bool value and then check it, which is a common pattern that can be nicely rewritten using separate function. Something like this (surely it's possible to come up with a better name for the function if you understand what the code does:)):
```cpp
bool InitializedResult(ComplexStrippedPartition& result, ...) {
   for (AttributeSet::size_type attr = attribute_set.find_first();
     attr != attribute_set.size(); attr = attribute_set.find_next(attr)) {
    AttributeSet one_less = deleteAttribute(attribute_set, attr);

    if (one_less.any() && cache_.Contains(one_less)) {
        result = cache_.Get(one_less);
        CallProduct(attr);
    }
}

 ComplexStrippedPartition GetStrippedPartition(AttributeSet const& attribute_set,
                                                  DataFrame const& data) {
    ...
    if (!InitializedResult(result, ...)) {
                    result = data.IsAttributesMostlyRangeBased(attribute_set)
                             ? ComplexStrippedPartition::Create<true>(data)
                             : ComplexStrippedPartition::Create<false>(data);

            for (AttributeSet::size_type attr = attribute_set.find_first();
                 attr != attribute_set.size(); attr = attribute_set.find_next(attr)) {
                CallProduct(attr);
            }
    }
}
```
Tbh I don't think it's very important, so up to you if you want to fix it

Reply:
`std::optional<ComplexStrippedPartition>` was also used to work around problems with the const reference to the `DataFrame` in `ComplexStrippedPartition` class (copy `DataFrame` in `operator=`, etc. ). I first [solved](https://github.com/Mstrutov/Desbordante/pull/355/commits/2597415c240e076b9cf11ccf9754d8387b6644ee) these old problems by using `std::shared_ptr<DataFrame>`. Next avoided using std::optional in cache.